### PR TITLE
[Java/C/C++] Concurrent publication fixes.

### PR DESCRIPTION
**Pull request by vyazelenko**
This PR contains the following changes:
- [Java/C/C++] Handle `term_id` moving forward between the position checks and an actual claim of the space in the buffer, i.e. instead of failing with error use the claimed space directly and return re-computed position based on the claimed `term_id/term_offset`.
- [Java/C/C++] Streamline checks for max position reached in concurrent publication, i.e. always consider the entire message frame length not just the payload length.
- [Java/C/C++/Wrapper] Use volatile load to get the latest tail of the next term.
- [C] Fix position computation to include the `term_offset` which was not considered before thus making C and Java client behave differently.
- [Java/C/C++/Wrapper] Remove `TermAppender/aeron_term_appender`.

### CODE REVIEW

**Comment by mikeb01** [comment]
Do we need to make any similar changes to the C++ API?

**Comment by vyazelenko** [comment]
@mikeb01 Yes, I'll look into it tomorrow.

File: aeron-client/src/main/cpp/concurrent/logbuffer/LogBufferDescriptor.h [hunk 1]
```diff
@@ -269,9 +270,17 @@ inline void endOfStreamPosition(AtomicBuffer &logMetaDataBuffer, std::int64_t po
     logMetaDataBuffer.putInt64Ordered(LOG_END_OF_STREAM_POSITION_OFFSET, position);
 }
 
+inline std::int32_t computeTermCount(
+    std::int32_t termId,
+    std::int32_t initialTermId) noexcept
+{
+    const std::int64_t difference = static_cast<std::int64_t>(termId) - static_cast<std::int64_t>(initialTermId);
+    return static_cast<std::int32_t>(difference & 0xFFFFFFFF);
+}
```
Comment (mikeb01) [inline aeron-client/src/main/cpp/concurrent/logbuffer/LogBufferDescriptor.h:279]: This change needs to be copied across to the wrapper.

**Review by mikeb01 — CHANGES_REQUESTED** [review]

File: aeron-client/src/main/cpp/concurrent/logbuffer/LogBufferDescriptor.h (hunk 1 above)
Comment (vyazelenko) [reply aeron-client/src/main/cpp/concurrent/logbuffer/LogBufferDescriptor.h:279]: Done but frankly speaking I don't understand why we have `LogBufferDescriptor.h` in the wrapper in the first place.

File: aeron-client/src/main/c/aeron_publication.c [hunk 2]
```diff
@@ -113,6 +112,368 @@ int aeron_publication_close(
...
+        header->frame_header.type = AERON_HDR_TYPE_PAD;
+        AERON_PUT_ORDERED(header->frame_header.frame_length, padding_length);
+    }
+
+    if (position >= publication->max_possible_position)
+    {
+        return AERON_PUBLICATION_MAX_POSITION_EXCEEDED;
+    }
+
+    aeron_logbuffer_rotate_log(publication->log_meta_data, term_id - publication->initial_term_id, term_id);
```
[the remark this answers is not in the document]
Comment (vyazelenko) [reply aeron-client/src/main/c/aeron_publication.c]: Done. I've missed the `term_count` checks.

File: aeron-client/src/main/c/aeron_publication.c [hunk 3]
```diff
@@ -139,33 +500,30 @@ int64_t aeron_publication_offer(
         const int64_t limit = aeron_counter_get_volatile(publication->position_limit);
         const int32_t term_count = aeron_logbuffer_active_term_count(publication->log_meta_data);
         const size_t index = aeron_logbuffer_index_by_term_count(term_count);
-        const int64_t raw_tail = aeron_term_appender_raw_tail_volatile(
+        const int64_t raw_tail = aeron_publication_raw_tail_volatile(
             &publication->log_meta_data->term_tail_counters[index]);
-        const int64_t term_offset = raw_tail & 0xFFFFFFFF;
+        const int32_t term_offset = (int32_t)(raw_tail & 0xFFFFFFFF);
```
[the remark this answers is not in the document]
Comment (vyazelenko) [reply aeron-client/src/main/c/aeron_publication.c:506]: It is deliberate here as we expect that when multiple threads race they will increment the `term_offset` pass the `term_length` in which case the log rotation will be triggered. Thus it is possible to observe the `term_offset > term_length` and we should use the exact value for position validation etc.

File: aeron-client/src/test/java/io/aeron/ConcurrentPublicationConcurrencyTest.java [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+    @AfterEach
+    void tearDown()
+    {
+        CloseHelper.closeAll(aeron, mediaDriver);
+    }
+
+    @Test
+    @InterruptAfter(300)
+    void test() throws InterruptedException
```
[the remark this answers is not in the document]
Comment (vyazelenko) [reply aeron-client/src/test/java/io/aeron/ConcurrentPublicationConcurrencyTest.java]: Done.

**Review by mikeb01 — APPROVED** [review]

**Comment by mikeb01** [comment]
LGTM

File: aeron-client/src/main/cpp/concurrent/logbuffer/ExclusiveTermAppender.h [hunk 5]
```diff
@@ -325,7 +324,7 @@ class ExclusiveTermAppender
             FrameDescriptor::frameLengthOrdered(termBuffer, termOffset, paddingLength);
         }
 
-        return TERM_APPENDER_FAILED;
+        return -2;
```
Comment (mjpt777) [inline aeron-client/src/main/cpp/concurrent/logbuffer/ExclusiveTermAppender.h:327]: Avoid use of magic number.

**Pull request merged**